OpenRA.Mods.Common/Traits/SupportPowers/SelectDirectionalTarget.cs

```diff
@@ -0,0 +1,188 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class SelectDirectionalTarget : IOrderGenerator
+	{
+		readonly string order;
+		readonly SupportPowerManager manager;
+		readonly string cursor;
+		readonly Animation targetCursor;
+
+		readonly string[] arrows = { "arrow-t", "arrow-tl", "arrow-l", "arrow-bl", "arrow-b", "arrow-br", "arrow-r", "arrow-tr" };
+		readonly Arrow[] directionArrows;
+
+		CPos targetCell;
+		int2 location;
+		int2 dragLocation;
+		bool beginDrag;
+		bool dragStarted;
+		Arrow currentArrow;
+
+		public SelectDirectionalTarget(World world, string order, SupportPowerManager manager, string cursor, string targetPlaceHolderCursorAnimation,
+			string directionCursorAnimation)
+		{
+			this.order = order;
+			this.manager = manager;
+			this.cursor = cursor;
+
+			targetCursor = new Animation(world, targetPlaceHolderCursorAnimation);
+			targetCursor.PlayRepeating("cursor");
+
+			directionArrows = LoadArrows(directionCursorAnimation, world, arrows.Length);
+		}
+
+		IEnumerable<Order> IOrderGenerator.Order(World world, CPos cell, int2 worldPixel, MouseInput mi)
+		{
+			if (mi.Button == MouseButton.Right)
+			{
+				world.CancelInputMode();
+				yield break;
+			}
+
+			if (mi.Button == MouseButton.Left && mi.Event == MouseInputEvent.Down)
+			{
+				targetCell = cell;
+				location = mi.Location;
+				beginDrag = true;
+
+				yield break;
+			}
+
+			if (mi.Event == MouseInputEvent.Move)
+			{
+				if (beginDrag)
+				{
+					dragLocation = mi.Location;
+					var angle = AngleBetween(location, dragLocation);
+					currentArrow = GetArrow(angle);
+					dragStarted = true;
+
+					yield break;
+				}
+			}
+
+			if (mi.Button == MouseButton.Left && mi.Event == MouseInputEvent.Up)
+			{
+				var ret = OrderInner().FirstOrDefault();
+				if (ret == null)
```

Comment:
It looks like this can never be null? Can we inline the contents of `OrderInner` here instead?